scipy/_lib/_util.py

```diff
@@ -647,3 +647,70 @@ def _nan_allsame(a, axis, keepdims=False):
             return np.full(shp, fill_value=True, dtype=bool)
     a0 = _first_nonnan(a, axis=axis)
     return ((a0 == a) | np.isnan(a)).all(axis=axis, keepdims=keepdims)
+
+
+def _rename_parameter(old_name, new_name, dep_version=None):
+    """
+    Generate decorator for backward-compatible keyword renaming.
+
+    Apply the decorator generated by `_rename_parameter` to functions with a
+    recently renamed parameter to maintain backward-compatibility.
+
+    After decoration, the function behaves as follows:
+    If only the new parameter is passed into the function, behave as usual.
+    If only the old parameter is passed into the function (as a keyword), raise
+    a DeprecationWarning if `dep_version` is provided, and behave as usual
+    otherwise.
+    If both old and new parameters are passed into the function, raise a
+    DeprecationWarning if `dep_version` is provided, and raise the appropriate
+    TypeError (function got multiple values for argument).
+
+    Parameters
+    ----------
+    old_name : str
+        Old name of parameter
+    new_name : str
+        New name of parameter
+    dep_version : str, optional
+        Version of SciPy in which old parameter was deprecated
+
+    Notes
+    -----
+    Untested with functions that accept *args. Probably won't work as written.
+
+    """
+    def decorator(fun):
+        @functools.wraps(fun)
+        def wrapper(*args, **kwds):
+
+            # Check for intersection between positional and keyword args
+            params = list(inspect.signature(fun).parameters)
+            d_args = dict(zip(params, args))
+            intersection = set(d_args) & set(kwds)
+            if intersection:
+                message = (f"{fun.__name__}() got multiple values "
+                           f"for argument '{list(intersection)[0]}'")
+                raise TypeError(message)
+
+            # Consolidate other positional and keyword args into `kwds`
+            kwds.update(d_args)
+
+            new_param = kwds.get(new_name, None)
+            got_new = new_param is not None
+            got_keyword_old = kwds.get(old_name, None) is not None
+
+            if got_keyword_old and dep_version:
+                message = (f"Use of keyword argument `{old_name}` is "
+                           f"deprecated and replaced by `{new_name}`.  "
+                           f"Support for `{old_name}` will be removed two "
+                           f"feature releases after SciPy {dep_version}.")
+                warnings.warn(message, DeprecationWarning, stacklevel=2)
+            if got_keyword_old and got_new:
+                message = (f"{fun.__name__}() got multiple values for "
+                           f"argument now known as `{new_name}`")
+                raise TypeError(message)
+
+            kwds[new_name] = kwds.pop(old_name, new_param)
+            return fun(**kwds)
```

Comment:
I think this can be written much simpler:

```python
            if old_name in kwargs:
                if new_name in kwargs:
                    raise TypeError(...)
                if dep_version:
                    warnings.warn(...)
                kwargs = dict(kwargs)
                kwargs[new_name] = kwargs.pop(old_name)
            return fun(*args, **kwargs)
```

Reply:
Probably right. I might have been borrowing too much from another project. I'll look into it.

Reply:
Yup, I was spreading my brain a little too thin. In this case, all the code to "Consolidate other positional and keyword args into `kwds`" was copy-pasted unnecessary. We can simply a little further without defining those variables. Apparently I forgot about the keyword `in` on 2/19, too.
Thanks big time @peterbell10!